server/src/main/java/org/elasticsearch/index/query/CombinedFieldsQueryBuilder.java

```diff
@@ -0,0 +1,433 @@
+    private static final ParseField ZERO_TERMS_QUERY_FIELD = new ParseField("zero_terms_query");
+
+    private final Object value;
+    private final Map<String, Float> fieldsAndBoosts;
+    private Operator operator = Operator.OR;
+    private String minimumShouldMatch;
+    private ZeroTermsQueryOption zeroTermsQuery = ZeroTermsQueryOption.NONE;
+    private boolean autoGenerateSynonymsPhraseQuery = true;
+
+    private static final ConstructingObjectParser<CombinedFieldsQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(NAME,
+        a -> new CombinedFieldsQueryBuilder(a[0]));
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), QUERY_FIELD);
+        PARSER.declareStringArray((builder, values) -> {
+            Map<String, Float> fieldsAndBoosts = QueryParserHelper.parseFieldsAndWeights(values);
+            builder.fields(fieldsAndBoosts);
+        }, FIELDS_FIELD);
+
+        PARSER.declareString(CombinedFieldsQueryBuilder::operator, Operator::fromString, OPERATOR_FIELD);
+        PARSER.declareString(CombinedFieldsQueryBuilder::minimumShouldMatch, MINIMUM_SHOULD_MATCH_FIELD);
+        PARSER.declareBoolean(CombinedFieldsQueryBuilder::autoGenerateSynonymsPhraseQuery, GENERATE_SYNONYMS_PHRASE_QUERY);
+        PARSER.declareString(CombinedFieldsQueryBuilder::zeroTermsQuery, value -> {
+            if ("none".equalsIgnoreCase(value)) {
+                return ZeroTermsQueryOption.NONE;
+            } else if ("all".equalsIgnoreCase(value)) {
+                return ZeroTermsQueryOption.ALL;
+            } else {
+                throw new IllegalArgumentException("Unsupported [" + ZERO_TERMS_QUERY_FIELD.getPreferredName() + "] value [" + value + "]");
+            }
+        }, ZERO_TERMS_QUERY_FIELD);
+
+        PARSER.declareFloat(CombinedFieldsQueryBuilder::boost, BOOST_FIELD);
+        PARSER.declareString(CombinedFieldsQueryBuilder::queryName, NAME_FIELD);
+    }
+
+    /**
+     * Constructs a new text query.
+     */
+    public CombinedFieldsQueryBuilder(Object value, String... fields) {
+        if (value == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires query value");
+        }
+        if (fields == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires field list");
+        }
+        this.value = value;
+        this.fieldsAndBoosts = new TreeMap<>();
+        for (String field : fields) {
+            field(field);
+        }
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public CombinedFieldsQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        value = in.readGenericValue();
+        int size = in.readVInt();
+        fieldsAndBoosts = new TreeMap<>();
+        for (int i = 0; i < size; i++) {
+            String field = in.readString();
+            float boost = in.readFloat();
+            checkNegativeBoost(boost);
```

Comment:
do we need to check for `boost < 1` or considering that the stream came from encoding a valid `CombinedFieldsQueryBuilder` these checks are not necessary?

Reply:
I copied this from `MultiMatchQueryBuilder` but I agree it doesn't seem necessary. Perhaps it was added there to handle the mixed-version case (where some older nodes could have negative boosts), but that doesn't apply to this new query.